### PR TITLE
google sheets improvements

### PR DESCRIPTION
**Pull request by denar90**
- goole sheets stuff moved to typescript and async/await
- added deps: [promisify](https://github.com/paulmillr/micro-promisify), [readline-sync](https://github.com/anseki/readline-sync)
- tests

### CODE REVIEW

File: lib/sheets/gsheets.ts [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+
+const google = require('googleapis');
+const promisify = require('micro-promisify');
+
+import { AuthorizeCredentials, GSheetsAppendResultsOptions, GSheetsValuesToAppend } from '../../types/types';
+
+async function getRange(auth: AuthorizeCredentials, range: number, spreadsheetId: string): Promise<any> {
+  try {
+    const sheets = google.sheets('v4');
+    const response = await promisify(sheets.spreadsheets.values.get)({
```
Comment (denar90) [inline lib/sheets/gsheets.ts]: @paulirish `await promisify` is it too much?

File: lib/outh/google-outh.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (denar90) [inline lib/outh/google-outh.ts]: I think this file will help us to authorize to gdrive for `timeline-viewer` integration.

File: lib/sheets/gsheets.ts (hunk 1 above)
Comment (paulirish) [inline lib/sheets/gsheets.ts]: Seems ok

File: lib/outh/google-outh.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+const path = require('path');
+const GoogleAuth = require('google-auth-library');
+const promisify = require('micro-promisify');
+const readlineSync = require('readline-sync');
+
+const { getMessage } = require('../utils/messages');
+
+import { AuthorizeCredentials } from '../../types/types';
+
+const fsReadFile = promisify(require('fs').readFile);
```
Comment (paulirish) [inline lib/outh/google-outh.ts]: readFileSync and writeFileSync means you can avoid the promisify/await on these.

Comment (denar90) [inline lib/outh/google-outh.ts]: Yeap, make sense.

**Comment by denar90** [comment]
@paulirish thanks. Should we involve Sam? eval(@samccone) 😛

File: lib/outh/google-outh.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+const self = globalAny.self || this;
+self.setImmediate = function(callback:any) {
+  Promise.resolve().then(_ => callback.apply(null, [...arguments].slice(1)));
+  return 0;
+};
+
+const EEXIST = 'EEXIST';
+
+class GoogleOuth {
+  scopes: Array<string>;
```
Comment (samccone) [inline lib/outh/google-outh.ts]: whatever props are private should be marked as such

File: lib/outh/google-outh.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  tokenDir: string;
+  tokenPath: string;
+
+  constructor() {
+    // If modifying these this.scopes, delete your previously saved credentials
+    // at ~/.credentials/sheets.googleapis.com-nodejs-pwmetrics.json
+    this.scopes = ['https://www.googleapis.com/auth/spreadsheets'];
```
Comment (samccone) [inline lib/outh/google-outh.ts]: extract these strings to file level consts, also you can do these assignments on lines 30 - 32

File: lib/outh/google-outh.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+    this.tokenDir = path.join((process.env.HOME || process.env.HOMEPATH || process.env.USERPROFILE), '/.credentials/');
+    this.tokenPath = path.join(this.tokenDir, 'sheets.googleapis.com-nodejs-pwmetrics.json');
+  }
+
+  async authenticate(clientSecret:AuthorizeCredentials): Promise<any> {
```
Comment (samccone) [inline lib/outh/google-outh.ts]: any? Don't we know the return type of authorize?

File: lib/outh/google-outh.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    try {
+      return await this.authorize(clientSecret);
+    } catch(error) {
+      throw new Error(error);
+    }
+  }
+
+  async authorize(credentials:AuthorizeCredentials): Promise<any> {
```
Comment (samccone) [inline lib/outh/google-outh.ts]: any ditto

File: lib/outh/google-outh.ts [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    try {
+      const token = await this.getToken();
+      oauth2Client.credentials = typeof token === 'string' ? JSON.parse(token) : token;
+      return oauth2Client;
+    } catch(error) {
+      return await this.getNewToken(oauth2Client);
+    }
+  }
+
+  async getToken() {
```
Comment (samccone) [inline lib/outh/google-outh.ts]: make sure to mark private fns as private

File: lib/outh/google-outh.ts [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+    return await fsReadFile(this.tokenPath, 'utf8');
+  }
+
+  async getNewToken(oauth2Client:any): Promise<any> {
```
Comment (samccone) [inline lib/outh/google-outh.ts]: any nit

File: lib/outh/google-outh.ts [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+    try {
+      const authUrl = oauth2Client.generateAuthUrl({
+        access_type: 'offline',
+        scope: this.scopes
+      });
+
+      console.log(getMessage('G_OUTH_WITH_URL', authUrl));
```
Comment (samccone) [inline lib/outh/google-outh.ts]: drop log?

**Comment by denar90** [comment]
@samccone thanks for the review 👍 
I requested review because I don't wanna build just something. I wanna build something good :)
@paulirish @samccone don't hesitate leave comments to my PR. Even if it's a nit picks) 

P.S. I'll solve conflicts.

File: lib/outh/google-outh.ts [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+        access_type: 'offline',
+        scope: SCOPES
+      });
+      const code: string = this.readline(authUrl);
+      const token: any = await this.getOauth2ClientToken(oauth2Client, code);
+      oauth2Client.credentials = token;
+      this.storeToken(token);
+      return oauth2Client;
+    } catch (error) {
+      throw getMessage('G_OUTH_ACCESS_ERROR',  error.message);
```
Comment (samccone) [inline lib/outh/google-outh.ts]: we should always be throwing exceptions and not strings

File: lib/outh/google-outh.ts [hunk 12]
```diff
@@ -0,0 +1,106 @@
...
+  private storeToken(token:string): void {
+    try {
+      fs.mkdirSync(this.tokenDir);
+    } catch (error) {
+      if (error.code !== EEXIST) {
+        throw error;
+      }
+    }
+    fs.writeFileSync(this.tokenPath, JSON.stringify(token));
+    console.log(getMessage('G_OUTH_STORED_TOKEN', this.tokenPath));
```
Comment (samccone) [inline lib/outh/google-outh.ts]: drop log?

File: lib/sheets/gsheets.ts [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+    return result.concat(value.slice(3).join('\t')).concat('\n');
+  }, []).join('');
+};
+
+async function appendResults(auth: Oauth2Client, valuesToAppend: Array<GSheetsValuesToAppend>, options:GSheetsAppendResultsOptions):Promise<any> {
+  try {
+    const sheets = google.sheets('v4');
+    // clone values to append
+    const values = Object.assign([], valuesToAppend);
+    console.log(getMessage('G_SHEETS_APPENDING', formatValues(valuesToAppend)));
```
Comment (samccone) [inline lib/sheets/gsheets.ts]: drop log(s) everywhere?

File: lib/sheets/index.ts [hunk 14]
```diff
@@ -44,12 +46,11 @@ class Sheets {
     }
   }
 
-  appendResultsToGSheets(results: Array<MetricsResults>): Promise<any[]> {
-    let valuesToAppend: Array<Object> = [];
+  async appendResultsToGSheets(results: Array<MetricsResults>): Promise<any> {
```
Comment (samccone) [inline lib/sheets/index.ts]: do we need the explicit return type? I think typescript can just pick this up

File: lib/outh/google-outh.ts [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+
+class GoogleOuth {
+  private tokenDir: string = path.join((process.env.HOME || process.env.HOMEPATH || process.env.USERPROFILE), '/.credentials/');
+  private tokenPath: string = path.join(this.tokenDir, 'sheets.googleapis.com-nodejs-pwmetrics.json');
+
+  async authenticate(clientSecret:AuthorizeCredentials): Promise<Oauth2Client> {
+    try {
+      return await this.authorize(clientSecret);
+    } catch(error) {
+      throw new Error(error);
```
Comment (denar90) [inline lib/outh/google-outh.ts]: Yeap...

File: lib/outh/google-outh.ts (hunk 12 above)
Comment (denar90) [inline lib/outh/google-outh.ts]: It's nice to see what results have been appended
```
Appending...
541.09  541.099 559.483 559.416 559.416 559.483 559.483

Appended
541.09  541.099 559.483 559.416 559.416 559.483 559.483

```

Wanna get rid of it or use kinda logger?

File: lib/sheets/gsheets.ts (hunk 13 above)
Comment (denar90) [inline lib/sheets/gsheets.ts]: Sorry, comment above related to this log.

**Comment by denar90** [comment]
@samccone thanks 🎉

**Pull request merged**